cpp/src/hash/sha_hash.cu

```diff
@@ -0,0 +1,695 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/hashing.hpp>
+#include <cudf/detail/utilities/hash_functions.cuh>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/strings/detail/utilities.cuh>
+#include <cudf/table/table_device_view.cuh>
+
+#include <hash/hash_constants.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+#include <type_traits>
+
+#include <thrust/fill.h>
+#include <thrust/for_each.h>
+#include <thrust/iterator/constant_iterator.h>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+// SHA supported leaf data type check
+bool sha_type_check(data_type dt)
+{
+  return !is_chrono(dt) && (is_fixed_width(dt) || (dt.id() == type_id::STRING));
+}
+
+CUDA_DEVICE_CALLABLE uint32_t rotate_bits_left(uint32_t x, int8_t r)
+{
+  // This function is equivalent to (x << r) | (x >> (32 - r))
+  return __funnelshift_l(x, x, r);
+}
+
+CUDA_DEVICE_CALLABLE uint32_t rotate_bits_right(uint32_t x, int8_t r)
+{
+  // This function is equivalent to (x >> r) | (x << (32 - r))
+  return __funnelshift_r(x, x, r);
+}
+
+CUDA_DEVICE_CALLABLE uint64_t rotate_bits_right(uint64_t x, int8_t r)
+{
+  return (x >> r) | (x << (64 - r));
+}
+
+// Swap the endianness of a 32 bit value
+CUDA_DEVICE_CALLABLE uint32_t swap_endian(uint32_t x)
+{
+  // The selector 0x0123 reverses the byte order
+  return __byte_perm(x, 0, 0x0123);
+}
+
+// Swap the endianness of a 64 bit value
+// There is no CUDA intrinsic for permuting bytes in 64 bit integers
+CUDA_DEVICE_CALLABLE uint64_t swap_endian(uint64_t x)
+{
+  // Reverse the endianness of each 32 bit section
+  uint32_t low_bits  = swap_endian(static_cast<uint32_t>(x));
+  uint32_t high_bits = swap_endian(static_cast<uint32_t>(x >> 32));
+  // Reassemble a 64 bit result, swapping the low bits and high bits
+  return (static_cast<uint64_t>(low_bits) << 32) | (static_cast<uint64_t>(high_bits));
+};
+
+}  // namespace
+
+template <typename HasherT>
+struct SHAHash {
+  /**
+   * @brief Execute SHA on input data chunks.
+   *
+   * This accepts arbitrary data, handles it as bytes, and calls the hash step
+   * when the buffer is filled up to message_chunk_size bytes.
+   */
+  template <typename TKey, typename Hasher = HasherT>
+  void CUDA_DEVICE_CALLABLE process(TKey const& key,
+                                    typename Hasher::sha_intermediate_data* hash_state) const
+  {
+    uint32_t const len  = sizeof(TKey);
+    uint8_t const* data = reinterpret_cast<uint8_t const*>(&key);
+    hash_state->message_length += len;
+
+    if (hash_state->buffer_length + len < Hasher::message_chunk_size) {
+      std::memcpy(hash_state->buffer + hash_state->buffer_length, data, len);
```

Comment:
What are the alignments of the source/destination here? 

As written, this is going to result in 1B load/stores. If you know the source/dest have an alignment greater than 1B, it would be much more performant to use larger load/stores.

Reply:
Discussed offline. The data is packed densely in the byte buffer, so we really can't leverage alignment greater than 1B.